### PR TITLE
Feature: Mouse suggestions

### PR DESCRIPTION
**Pull request by LinqLover**
This PR attempts to improve overall mouse support in connection with block suggestions.

Miscellaneous things that still could be improved but are not directly related to this PR:

- No suggestions for `SBStNumber`s (this would be helpful if you wanted to switch to a different block, which currently still requires the use of your keyboard)
- Suggestions do not disappear if another (maybe non-sandblockish) editor gains the focus. I tried to change this in `keyboardFocusChanged:`, but then the suggestions would disappear much too often. Maybe we should try to explicitly detect concurring editors when they are focused? But there is no appropriate interface for this.

### CODE REVIEW

File: packages/Sandblocks-Core/SBEditor.class.st [hunk 1]
```diff
@@ -28,7 +28,8 @@ Class {
 		'dragHelper',
 		'cursor',
 		'activateForcesButton',
-		'hoverHighlight'
+		'hoverHighlights',
```
Comment (LinqLover) [inline packages/Sandblocks-Core/SBEditor.class.st:31]:
```suggestion
		'hoverHighlight',
```

File: packages/Sandblocks-Core/SBSuggestionItem.class.st [hunk 2]
```diff
@@ -72,6 +72,18 @@ SBSuggestionItem >> drawOn: aCanvas [
 		color: (foreground alpha: 0.5)
 ]
 
+{ #category : #'as yet unclassified' }
```
Comment (LinqLover) [inline packages/Sandblocks-Core/SBSuggestionItem.class.st:75]: recategorize

File: packages/Sandblocks-Core/SBSuggestionItem.class.st [hunk 3]
```diff
@@ -108,6 +126,18 @@ SBSuggestionItem >> minExtent [
 	^ (font widthOfString: selector) @ font height + (self layoutInset * 2) + (self cellGap @ 0) + ((font widthOfString: label) @ 0)
 ]
 
+{ #category : #'as yet unclassified' }
```
Comment (LinqLover) [inline packages/Sandblocks-Core/SBSuggestionItem.class.st:129]: recategorize

File: packages/Sandblocks-Core/SBSuggestionMenu.class.st [hunk 4]
```diff
@@ -1,19 +1,34 @@
...
-	#superclass : #Morph,
+	#superclass : #ScrollPane,
 	#instVars : [
 		'activeIndex',
-		'editor'
+		'editor',
+		'container'
 	],
-	#category : 'Sandblocks-Core'
+	#category : #'Sandblocks-Core'
```
Comment (LinqLover) [inline packages/Sandblocks-Core/SBSuggestionMenu.class.st:9]: This is my first experience with Tonel. :D How could this happen?

File: packages/Sandblocks-Core/SBSuggestionMenu.class.st [hunk 5]
```diff
@@ -1,19 +1,34 @@
...
 }
 
+{ #category : #'as yet unclassified' }
```
Comment (LinqLover) [inline packages/Sandblocks-Core/SBSuggestionMenu.class.st:12]: recategorize

File: packages/Sandblocks-Smalltalk/SBStASTNode.class.st [hunk 6]
```diff
@@ -247,6 +247,18 @@ SBStASTNode >> sourceString [
 	^ String streamContents: [:s | self writeSourceOn: s]
 ]
 
+{ #category : #'as yet unclassified' }
```
Comment (LinqLover) [inline packages/Sandblocks-Smalltalk/SBStASTNode.class.st:250]: recategorize

File: packages/Sandblocks-Smalltalk/SBStSignature.class.st [hunk 7]
```diff
@@ -42,6 +42,12 @@ SBStSignature >> cellGap [
 	^ self containingSandblock colorPolicy messageCellGap
 ]
 
+{ #category : #'as yet unclassified' }
```
Comment (LinqLover) [inline packages/Sandblocks-Smalltalk/SBStSignature.class.st:45]: recategorize

**Comment by LinqLover** [comment]
Possible layout improvement:

image

In this example, the suggestion menu could be shrunk.

File: packages/Sandblocks-Core/SBBlock.class.st [hunk 8]
```diff
@@ -421,16 +422,24 @@ SBBlock >> completeSubtree [
...
+							class selectors sorted do: [:sel |
+								(sel asString sandblockMatch: aString) ifTrue: [stream nextPut: sel asString]]]].
+					allMatches := Array streamContents: [:stream |
 						baseClass withAllSuperclassesDo: [:class |
-							class methodsDo: [:m | | sel |
-								sel := m selector asString.
-								(sel sandblockMatch: aString) ifTrue: [stream nextPut: sel]]]]]
+							class selectors sorted do: [:sel |
+								stream nextPut: sel asString]]].
+					(exactMatches, fuzzyMatches, allMatches) withoutDuplicates]
```
Comment (LinqLover) [inline packages/Sandblocks-Core/SBBlock.class.st:442]: I'm currently reworking this part again to add support for class-specific suggestion filters (but will take some time to submit another PR) - but eventually, the duplication is planned to be eliminated definitively. :-)

**Comment by tom95** [comment]
Very nice stuff, thank you! I just went ahead and merged and then pushed a tiny commit right after that addresses the layout improvement you had mentioned (among two other things): https://github.com/tom95/sandblocks/commit/d47b6b61784ecefeb96862ac73443c2fc99d0fea

Being able to click on suggestion items feels oddly satisfying, I'm quite surprised I hadn't seen this need myself yet :D
